In [2]:
import pandas as pd

In [3]:
pd.DataFrame({'Yes':[13,150],'No':[151,12]})

,Yes,No
0,13,151
1,150,12


In [4]:
pd.DataFrame({'Bob':["I liked it","It's awful"],'Alex':["Pretty Good","Bland"]},index=["Dish1","Dish2"])

,Bob,Alex
Dish1,I liked it,Pretty Good
Dish2,It's awful,Bland


In [5]:
pd.Series([35,34,31,37],index=["2015","2016","2017","2018"],name="Product A")

2015    35
2016    34
2017    31
2018    37
Name: Product A, dtype: int64

In [6]:
medical_data=pd.read_excel("/home/sasig/Documents/Pract/Teleflex/Forecast_Accuracy_Assessment_Brief_v2/Medical_Device_Forecast_Data.xlsx")

In [7]:
medical_data.shape

(1000, 12)

In [8]:
medical_data.head()

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC
0,SKU-100,Orthopedic Screws,APAC,2024-11,116,332,540,15,No,No,1.0,332.0
1,SKU-100,Infusion Pumps,North America,2024-07,269,127,291,5,No,No,1.0,127.0
2,SKU-100,Orthopedic Screws,LATAM,2025-02,512,258,523,0,Yes,Yes,1.3,335.4
3,SKU-100,Orthopedic Screws,APAC,2025-08,474,158,399,0,No,Yes,1.0,158.0
4,SKU-100,Orthopedic Screws,APAC,2024-04,563,262,614,0,Yes,Yes,1.3,340.6


In [9]:
medical_data

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC
0,SKU-100,Orthopedic Screws,APAC,2024-11,116,332,540,15,No,No,1.0,332.0
1,SKU-100,Infusion Pumps,North America,2024-07,269,127,291,5,No,No,1.0,127.0
2,SKU-100,Orthopedic Screws,LATAM,2025-02,512,258,523,0,Yes,Yes,1.3,335.4
3,SKU-100,Orthopedic Screws,APAC,2025-08,474,158,399,0,No,Yes,1.0,158.0
4,SKU-100,Orthopedic Screws,APAC,2024-04,563,262,614,0,Yes,Yes,1.3,340.6
...,...,...,...,...,...,...,...,...,...,...,...,...
995,SKU-109,MRI Coils,North America,2024-05,10,23,54,15,No,No,1.0,23.0
996,SKU-109,Infusion Pumps,LATAM,2024-11,12,24,56,15,No,No,1.0,24.0
997,SKU-109,Surgical Lasers,APAC,2025-11,0,20,5,0,Yes,Yes,1.3,26.0
998,SKU-109,Infusion Pumps,EMEA,2024-03,186,90,156,0,No,Yes,1.0,90.0


In [10]:
medical_data.Region

0               APAC
1      North America
2              LATAM
3               APAC
4               APAC
           ...      
995    North America
996            LATAM
997             APAC
998             EMEA
999    North America
Name: Region, Length: 1000, dtype: str

In [11]:
medical_data["Region"]

0               APAC
1      North America
2              LATAM
3               APAC
4               APAC
           ...      
995    North America
996            LATAM
997             APAC
998             EMEA
999    North America
Name: Region, Length: 1000, dtype: str

In [12]:
medical_data["Region"][0]

'APAC'

In [13]:
medical_data.iloc[0]

Product_ID                         SKU-100
Category                 Orthopedic Screws
Region                                APAC
Month                              2024-11
Actual_Sales                           116
ERP_Stat_FC                            332
Consensus_FC                           540
Stockout_Days                           15
Promotion_Flag                          No
In_Stock_Flag                           No
Promo_Uplift_Factor                    1.0
Bias_Adjusted_Stat_FC                332.0
Name: 0, dtype: object

Both `loc` and `iloc` are row-first, column-second. This is the opposite of what we do in native Python, which is column-first, row-second.

In [14]:
medical_data.iloc[:,2]

0               APAC
1      North America
2              LATAM
3               APAC
4               APAC
           ...      
995    North America
996            LATAM
997             APAC
998             EMEA
999    North America
Name: Region, Length: 1000, dtype: str

In [15]:
medical_data.iloc[2:5,2]

2    LATAM
3     APAC
4     APAC
Name: Region, dtype: str

In [16]:
medical_data.iloc[[1,3,7],2]

1    North America
3             APAC
7    North America
Name: Region, dtype: str

In [17]:
medical_data.iloc[-5:]

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC
995,SKU-109,MRI Coils,North America,2024-05,10,23,54,15,No,No,1.0,23.0
996,SKU-109,Infusion Pumps,LATAM,2024-11,12,24,56,15,No,No,1.0,24.0
997,SKU-109,Surgical Lasers,APAC,2025-11,0,20,5,0,Yes,Yes,1.3,26.0
998,SKU-109,Infusion Pumps,EMEA,2024-03,186,90,156,0,No,Yes,1.0,90.0
999,SKU-109,Surgical Lasers,North America,2025-02,0,20,5,20,No,No,1.0,20.0


In [18]:
medical_data.loc[0,"Region"]

'APAC'

In `iloc` we use indices of rows/columns, but in `loc` we use the names of the rows/columns

In [19]:
medical_data.loc[:,["Actual_Sales","ERP_Stat_FC","Bias_Adjusted_Stat_FC"]]

,Actual_Sales,ERP_Stat_FC,Bias_Adjusted_Stat_FC
0,116,332,332.0
1,269,127,127.0
2,512,258,335.4
3,474,158,158.0
4,563,262,340.6
...,...,...,...
995,10,23,23.0
996,12,24,24.0
997,0,20,26.0
998,186,90,90.0


In this case `df.iloc[0:1000]` will return 1000 entries, while `df.loc[0:1000]` return 1001 of them! To get 1000 elements using loc, you will need to go one lower and ask for `df.loc[0:999]`

In [20]:
medical_data.set_index("Month")

,Product_ID,Category,Region,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC
Month,,,,,,,,,,,
2024-11,SKU-100,Orthopedic Screws,APAC,116,332,540,15,No,No,1.0,332.0
2024-07,SKU-100,Infusion Pumps,North America,269,127,291,5,No,No,1.0,127.0
2025-02,SKU-100,Orthopedic Screws,LATAM,512,258,523,0,Yes,Yes,1.3,335.4
2025-08,SKU-100,Orthopedic Screws,APAC,474,158,399,0,No,Yes,1.0,158.0
2024-04,SKU-100,Orthopedic Screws,APAC,563,262,614,0,Yes,Yes,1.3,340.6
...,...,...,...,...,...,...,...,...,...,...,...
2024-05,SKU-109,MRI Coils,North America,10,23,54,15,No,No,1.0,23.0
2024-11,SKU-109,Infusion Pumps,LATAM,12,24,56,15,No,No,1.0,24.0
2025-11,SKU-109,Surgical Lasers,APAC,0,20,5,0,Yes,Yes,1.3,26.0


In [21]:
medical_data.head()

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC
0,SKU-100,Orthopedic Screws,APAC,2024-11,116,332,540,15,No,No,1.0,332.0
1,SKU-100,Infusion Pumps,North America,2024-07,269,127,291,5,No,No,1.0,127.0
2,SKU-100,Orthopedic Screws,LATAM,2025-02,512,258,523,0,Yes,Yes,1.3,335.4
3,SKU-100,Orthopedic Screws,APAC,2025-08,474,158,399,0,No,Yes,1.0,158.0
4,SKU-100,Orthopedic Screws,APAC,2024-04,563,262,614,0,Yes,Yes,1.3,340.6


In [22]:
medical_data.In_Stock_Flag=="Yes"

0      False
1      False
2       True
3       True
4       True
       ...  
995    False
996    False
997     True
998     True
999    False
Name: In_Stock_Flag, Length: 1000, dtype: bool

In [23]:
medical_data.loc[medical_data.In_Stock_Flag=="Yes"]

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC
2,SKU-100,Orthopedic Screws,LATAM,2025-02,512,258,523,0,Yes,Yes,1.3,335.4
3,SKU-100,Orthopedic Screws,APAC,2025-08,474,158,399,0,No,Yes,1.0,158.0
4,SKU-100,Orthopedic Screws,APAC,2024-04,563,262,614,0,Yes,Yes,1.3,340.6
5,SKU-100,Surgical Lasers,EMEA,2024-07,50,16,54,0,No,Yes,1.0,16.0
7,SKU-100,MRI Coils,North America,2024-11,75,31,81,0,No,Yes,1.0,31.0
...,...,...,...,...,...,...,...,...,...,...,...,...
990,SKU-109,MRI Coils,APAC,2024-01,158,85,137,0,No,Yes,1.0,85.0
991,SKU-109,Surgical Lasers,LATAM,2024-09,0,20,5,0,No,Yes,1.0,20.0
994,SKU-109,Orthopedic Screws,LATAM,2024-03,504,176,510,0,No,Yes,1.0,176.0
997,SKU-109,Surgical Lasers,APAC,2025-11,0,20,5,0,Yes,Yes,1.3,26.0


In [24]:
medical_data.loc[(medical_data.In_Stock_Flag=="Yes") & (medical_data.Actual_Sales>=200)]

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC
2,SKU-100,Orthopedic Screws,LATAM,2025-02,512,258,523,0,Yes,Yes,1.3,335.4
3,SKU-100,Orthopedic Screws,APAC,2025-08,474,158,399,0,No,Yes,1.0,158.0
4,SKU-100,Orthopedic Screws,APAC,2024-04,563,262,614,0,Yes,Yes,1.3,340.6
12,SKU-100,MRI Coils,North America,2024-01,215,99,200,0,No,Yes,1.0,99.0
17,SKU-100,Infusion Pumps,North America,2024-11,269,142,249,0,Yes,Yes,1.3,184.6
...,...,...,...,...,...,...,...,...,...,...,...,...
966,SKU-109,Stents (Cardiac),North America,2025-08,238,116,260,0,No,Yes,1.0,116.0
972,SKU-109,MRI Coils,APAC,2025-10,220,96,225,0,No,Yes,1.0,96.0
977,SKU-109,MRI Coils,North America,2024-03,268,97,221,0,No,Yes,1.0,97.0
985,SKU-109,Orthopedic Screws,EMEA,2024-05,571,210,526,0,No,Yes,1.0,210.0


In [25]:
medical_data.loc[(medical_data.In_Stock_Flag=="Yes") | (medical_data.Actual_Sales>=200)]

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC
1,SKU-100,Infusion Pumps,North America,2024-07,269,127,291,5,No,No,1.0,127.0
2,SKU-100,Orthopedic Screws,LATAM,2025-02,512,258,523,0,Yes,Yes,1.3,335.4
3,SKU-100,Orthopedic Screws,APAC,2025-08,474,158,399,0,No,Yes,1.0,158.0
4,SKU-100,Orthopedic Screws,APAC,2024-04,563,262,614,0,Yes,Yes,1.3,340.6
5,SKU-100,Surgical Lasers,EMEA,2024-07,50,16,54,0,No,Yes,1.0,16.0
...,...,...,...,...,...,...,...,...,...,...,...,...
990,SKU-109,MRI Coils,APAC,2024-01,158,85,137,0,No,Yes,1.0,85.0
991,SKU-109,Surgical Lasers,LATAM,2024-09,0,20,5,0,No,Yes,1.0,20.0
994,SKU-109,Orthopedic Screws,LATAM,2024-03,504,176,510,0,No,Yes,1.0,176.0
997,SKU-109,Surgical Lasers,APAC,2025-11,0,20,5,0,Yes,Yes,1.3,26.0


In [26]:
medical_data.loc[medical_data.Region.isin(["North America","APAC"])]

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC
0,SKU-100,Orthopedic Screws,APAC,2024-11,116,332,540,15,No,No,1.0,332.0
1,SKU-100,Infusion Pumps,North America,2024-07,269,127,291,5,No,No,1.0,127.0
3,SKU-100,Orthopedic Screws,APAC,2025-08,474,158,399,0,No,Yes,1.0,158.0
4,SKU-100,Orthopedic Screws,APAC,2024-04,563,262,614,0,Yes,Yes,1.3,340.6
6,SKU-100,Infusion Pumps,APAC,2024-10,31,55,145,20,No,No,1.0,55.0
...,...,...,...,...,...,...,...,...,...,...,...,...
990,SKU-109,MRI Coils,APAC,2024-01,158,85,137,0,No,Yes,1.0,85.0
993,SKU-109,Infusion Pumps,APAC,2024-06,102,34,93,5,No,No,1.0,34.0
995,SKU-109,MRI Coils,North America,2024-05,10,23,54,15,No,No,1.0,23.0
997,SKU-109,Surgical Lasers,APAC,2025-11,0,20,5,0,Yes,Yes,1.3,26.0


In [27]:
medical_data.loc[medical_data.Actual_Sales.notnull()]

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC
0,SKU-100,Orthopedic Screws,APAC,2024-11,116,332,540,15,No,No,1.0,332.0
1,SKU-100,Infusion Pumps,North America,2024-07,269,127,291,5,No,No,1.0,127.0
2,SKU-100,Orthopedic Screws,LATAM,2025-02,512,258,523,0,Yes,Yes,1.3,335.4
3,SKU-100,Orthopedic Screws,APAC,2025-08,474,158,399,0,No,Yes,1.0,158.0
4,SKU-100,Orthopedic Screws,APAC,2024-04,563,262,614,0,Yes,Yes,1.3,340.6
...,...,...,...,...,...,...,...,...,...,...,...,...
995,SKU-109,MRI Coils,North America,2024-05,10,23,54,15,No,No,1.0,23.0
996,SKU-109,Infusion Pumps,LATAM,2024-11,12,24,56,15,No,No,1.0,24.0
997,SKU-109,Surgical Lasers,APAC,2025-11,0,20,5,0,Yes,Yes,1.3,26.0
998,SKU-109,Infusion Pumps,EMEA,2024-03,186,90,156,0,No,Yes,1.0,90.0


In [28]:
medical_data["Index_Backwards"]=range(len(medical_data)-1,-1,-1)

In [29]:
medical_data

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
0,SKU-100,Orthopedic Screws,APAC,2024-11,116,332,540,15,No,No,1.0,332.0,999
1,SKU-100,Infusion Pumps,North America,2024-07,269,127,291,5,No,No,1.0,127.0,998
2,SKU-100,Orthopedic Screws,LATAM,2025-02,512,258,523,0,Yes,Yes,1.3,335.4,997
3,SKU-100,Orthopedic Screws,APAC,2025-08,474,158,399,0,No,Yes,1.0,158.0,996
4,SKU-100,Orthopedic Screws,APAC,2024-04,563,262,614,0,Yes,Yes,1.3,340.6,995
...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,SKU-109,MRI Coils,North America,2024-05,10,23,54,15,No,No,1.0,23.0,4
996,SKU-109,Infusion Pumps,LATAM,2024-11,12,24,56,15,No,No,1.0,24.0,3
997,SKU-109,Surgical Lasers,APAC,2025-11,0,20,5,0,Yes,Yes,1.3,26.0,2
998,SKU-109,Infusion Pumps,EMEA,2024-03,186,90,156,0,No,Yes,1.0,90.0,1


In [30]:
medical_data.Actual_Sales.describe()

count    1000.000000
mean      151.446000
std       158.474355
min         0.000000
25%        32.000000
50%       101.000000
75%       219.250000
max       599.000000
Name: Actual_Sales, dtype: float64

In [31]:
medical_data.Region.describe()

count              1000
unique                4
top       North America
freq                272
Name: Region, dtype: object

In [32]:
medical_data.Product_ID.describe()

count        1000
unique         10
top       SKU-105
freq          110
Name: Product_ID, dtype: object

In [33]:
medical_data.Actual_Sales.mean()

np.float64(151.446)

In [34]:
medical_data.Region.unique()

<StringArray>
['APAC', 'North America', 'LATAM', 'EMEA']
Length: 4, dtype: str

In [35]:
medical_data.Category.unique()

<StringArray>
['Orthopedic Screws',    'Infusion Pumps',   'Surgical Lasers',
         'MRI Coils',  'Stents (Cardiac)']
Length: 5, dtype: str

In [36]:
medical_data.Category.value_counts()

Category
Surgical Lasers      210
Orthopedic Screws    202
MRI Coils            202
Infusion Pumps       194
Stents (Cardiac)     192
Name: count, dtype: int64

In [37]:
medical_actual_sales_mean=medical_data.Actual_Sales.mean()

In [38]:
medical_data.Actual_Sales.map(lambda p: p-medical_actual_sales_mean)

0      -35.446
1      117.554
2      360.554
3      322.554
4      411.554
        ...   
995   -141.446
996   -139.446
997   -151.446
998     34.554
999   -151.446
Name: Actual_Sales, Length: 1000, dtype: float64

In [39]:
def remean_actual_sales(row):
    row.Actual_Sales=row.Actual_Sales-medical_actual_sales_mean
    return row
medical_data.apply(remean_actual_sales,axis='columns')

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
0,SKU-100,Orthopedic Screws,APAC,2024-11,-35.446,332,540,15,No,No,1.0,332.0,999
1,SKU-100,Infusion Pumps,North America,2024-07,117.554,127,291,5,No,No,1.0,127.0,998
2,SKU-100,Orthopedic Screws,LATAM,2025-02,360.554,258,523,0,Yes,Yes,1.3,335.4,997
3,SKU-100,Orthopedic Screws,APAC,2025-08,322.554,158,399,0,No,Yes,1.0,158.0,996
4,SKU-100,Orthopedic Screws,APAC,2024-04,411.554,262,614,0,Yes,Yes,1.3,340.6,995
...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,SKU-109,MRI Coils,North America,2024-05,-141.446,23,54,15,No,No,1.0,23.0,4
996,SKU-109,Infusion Pumps,LATAM,2024-11,-139.446,24,56,15,No,No,1.0,24.0,3
997,SKU-109,Surgical Lasers,APAC,2025-11,-151.446,20,5,0,Yes,Yes,1.3,26.0,2
998,SKU-109,Infusion Pumps,EMEA,2024-03,34.554,90,156,0,No,Yes,1.0,90.0,1


In [40]:
medical_data.Actual_Sales.idxmax()

770

In [41]:
medical_data.groupby("Category").Category.count()

Category
Infusion Pumps       194
MRI Coils            202
Orthopedic Screws    202
Stents (Cardiac)     192
Surgical Lasers      210
Name: Category, dtype: int64

In [42]:
medical_data.groupby("Category").Actual_Sales.min()

Category
Infusion Pumps       10
MRI Coils            10
Orthopedic Screws    80
Stents (Cardiac)     12
Surgical Lasers       0
Name: Actual_Sales, dtype: int64

In [43]:
medical_data.groupby(["Category"]).Actual_Sales.max()

Category
Infusion Pumps       299
MRI Coils            297
Orthopedic Screws    599
Stents (Cardiac)     297
Surgical Lasers      120
Name: Actual_Sales, dtype: int64

In [44]:
medical_data.groupby(['Category','Region']).Actual_Sales.mean()

Category           Region       
Infusion Pumps     APAC             110.058824
                   EMEA             133.116279
                   LATAM            128.482143
                   North America    118.868852
MRI Coils          APAC             110.355556
                   EMEA             123.230769
                   LATAM             88.190476
                   North America    134.020000
Orthopedic Screws  APAC             387.106383
                   EMEA             384.340000
                   LATAM            369.890909
                   North America    298.000000
Stents (Cardiac)   APAC             152.250000
                   EMEA             133.139535
                   LATAM            139.978261
                   North America    125.163636
Surgical Lasers    APAC              27.446809
                   EMEA              25.254902
                   LATAM             23.428571
                   North America     22.357143
Name: Actual_Sales, dtype: 

In [45]:
medical_groupby=medical_data.groupby(['Category','Region']).apply(lambda df: df.loc[df.Actual_Sales.idxmax()])
medical_groupby

Product_ID    Month  Actual_Sales  \
Category          Region                                            
Infusion Pumps    APAC             SKU-100  2025-08           244   
                  EMEA             SKU-105  2025-11           290   
                  LATAM            SKU-107  2025-07           299   
                  North America    SKU-105  2024-01           294   
MRI Coils         APAC             SKU-103  2024-07           297   
                  EMEA             SKU-103  2024-11           292   
                  LATAM            SKU-105  2024-01           267   
                  North America    SKU-100  2024-05           292   
Orthopedic Screws APAC             SKU-107  2024-01           599   
                  EMEA             SKU-100  2024-10           587   
                  LATAM            SKU-102  2024-05           592   
                  North America    SKU-106  2025-01           596   
Stents (Cardiac)  APAC             SKU-104  2025-06           290   
                  EMEA             SKU-107  2025-02           297   
                  LATAM            SKU-106  2024-03           296   
                  North America    SKU-102  2024-06           279   
Surgical Lasers   APAC             SKU-102  2024-01           120   
                  EMEA             SKU-101  2025-06           120   
                  LATAM            SKU-101  2024-09           120   
                  North America    SKU-100  2024-08           120   

                                 ERP_Stat_FC  Consensus_FC  Stockout_Days  \
Category          Region                                                    
Infusion Pumps    APAC                   132           245              0   
                  EMEA                   167           275              0   
                  LATAM                  119           285              0   
                  North America          143           249              0   
MRI Coils         APAC                   105           245              5   
                  EMEA                   130           278              0   
                  LATAM                  138           218              0   
                  North America          108           247              0   
Orthopedic Screws APAC                   301           524              0   
                  EMEA                   242           470              0   
                  LATAM                  180           577              0   
                  North America          196           508              0   
Stents (Cardiac)  APAC                   103           279              0   
                  EMEA                   153           310              0   
                  LATAM                  111           306              0   
                  North America          149           260              0   
Surgical Lasers   APAC                    49           127              5   
                  EMEA                    57           118              0   
                  LATAM                   64           107              0   
                  North America           40           110              0   

                                Promotion_Flag In_Stock_Flag  \
Category          Region                                       
Infusion Pumps    APAC                      No           Yes   
                  EMEA                      No           Yes   
                  LATAM                     No           Yes   
                  North America             No           Yes   
MRI Coils         APAC                      No            No   
                  EMEA                      No           Yes   
                  LATAM                     No           Yes   
                  North America             No           Yes   
Orthopedic Screws APAC                      No           Yes   
                  EMEA                      No           Yes   
                  LATAM                     No    

In [46]:
medical_data.groupby(['Category']).Actual_Sales.agg([len,min,max])

,len,min,max
Category,,,
Infusion Pumps,194,10,299
MRI Coils,202,10,297
Orthopedic Screws,202,80,599
Stents (Cardiac),192,12,297
Surgical Lasers,210,0,120


In [47]:
mi=medical_groupby.index

In [48]:
type(mi)

pandas.MultiIndex

In [49]:
medical_groupby=medical_groupby.reset_index()
medical_groupby

,Category,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
0,Infusion Pumps,APAC,SKU-100,2025-08,244,132,245,0,No,Yes,1.0,132.0,945
1,Infusion Pumps,EMEA,SKU-105,2025-11,290,167,275,0,No,Yes,1.0,167.0,459
2,Infusion Pumps,LATAM,SKU-107,2025-07,299,119,285,0,No,Yes,1.0,119.0,223
3,Infusion Pumps,North America,SKU-105,2024-01,294,143,249,0,No,Yes,1.0,143.0,445
4,MRI Coils,APAC,SKU-103,2024-07,297,105,245,5,No,No,1.0,105.0,665
5,MRI Coils,EMEA,SKU-103,2024-11,292,130,278,0,No,Yes,1.0,130.0,653
6,MRI Coils,LATAM,SKU-105,2024-01,267,138,218,0,No,Yes,1.0,138.0,494
7,MRI Coils,North America,SKU-100,2024-05,292,108,247,0,No,Yes,1.0,108.0,900
8,Orthopedic Screws,APAC,SKU-107,2024-01,599,301,524,0,No,Yes,1.0,301.0,229
9,Orthopedic Screws,EMEA,SKU-100,2024-10,587,242,470,0,No,Yes,1.0,242.0,958


In [50]:
medical_groupby.sort_values(by='Actual_Sales')

,Category,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
19,Surgical Lasers,North America,SKU-100,2024-08,120,40,110,0,No,Yes,1.0,40.0,915
18,Surgical Lasers,LATAM,SKU-101,2024-09,120,64,107,0,No,Yes,1.0,64.0,858
17,Surgical Lasers,EMEA,SKU-101,2025-06,120,57,118,0,Yes,Yes,1.3,74.1,886
16,Surgical Lasers,APAC,SKU-102,2024-01,120,49,127,5,No,No,1.0,49.0,722
0,Infusion Pumps,APAC,SKU-100,2025-08,244,132,245,0,No,Yes,1.0,132.0,945
6,MRI Coils,LATAM,SKU-105,2024-01,267,138,218,0,No,Yes,1.0,138.0,494
15,Stents (Cardiac),North America,SKU-102,2024-06,279,149,260,0,No,Yes,1.0,149.0,759
1,Infusion Pumps,EMEA,SKU-105,2025-11,290,167,275,0,No,Yes,1.0,167.0,459
12,Stents (Cardiac),APAC,SKU-104,2025-06,290,103,279,0,No,Yes,1.0,103.0,594
5,MRI Coils,EMEA,SKU-103,2024-11,292,130,278,0,No,Yes,1.0,130.0,653


In [51]:
medical_groupby.sort_values(by='Actual_Sales',ascending=False)

,Category,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
8,Orthopedic Screws,APAC,SKU-107,2024-01,599,301,524,0,No,Yes,1.0,301.0,229
11,Orthopedic Screws,North America,SKU-106,2025-01,596,196,508,0,No,Yes,1.0,196.0,334
10,Orthopedic Screws,LATAM,SKU-102,2024-05,592,180,577,0,No,Yes,1.0,180.0,749
9,Orthopedic Screws,EMEA,SKU-100,2024-10,587,242,470,0,No,Yes,1.0,242.0,958
2,Infusion Pumps,LATAM,SKU-107,2025-07,299,119,285,0,No,Yes,1.0,119.0,223
4,MRI Coils,APAC,SKU-103,2024-07,297,105,245,5,No,No,1.0,105.0,665
13,Stents (Cardiac),EMEA,SKU-107,2025-02,297,153,310,0,No,Yes,1.0,153.0,292
14,Stents (Cardiac),LATAM,SKU-106,2024-03,296,111,306,0,No,Yes,1.0,111.0,345
3,Infusion Pumps,North America,SKU-105,2024-01,294,143,249,0,No,Yes,1.0,143.0,445
5,MRI Coils,EMEA,SKU-103,2024-11,292,130,278,0,No,Yes,1.0,130.0,653


In [52]:
medical_groupby.sort_values(['Category','Actual_Sales'])

,Category,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
0,Infusion Pumps,APAC,SKU-100,2025-08,244,132,245,0,No,Yes,1.0,132.0,945
1,Infusion Pumps,EMEA,SKU-105,2025-11,290,167,275,0,No,Yes,1.0,167.0,459
3,Infusion Pumps,North America,SKU-105,2024-01,294,143,249,0,No,Yes,1.0,143.0,445
2,Infusion Pumps,LATAM,SKU-107,2025-07,299,119,285,0,No,Yes,1.0,119.0,223
6,MRI Coils,LATAM,SKU-105,2024-01,267,138,218,0,No,Yes,1.0,138.0,494
5,MRI Coils,EMEA,SKU-103,2024-11,292,130,278,0,No,Yes,1.0,130.0,653
7,MRI Coils,North America,SKU-100,2024-05,292,108,247,0,No,Yes,1.0,108.0,900
4,MRI Coils,APAC,SKU-103,2024-07,297,105,245,5,No,No,1.0,105.0,665
9,Orthopedic Screws,EMEA,SKU-100,2024-10,587,242,470,0,No,Yes,1.0,242.0,958
10,Orthopedic Screws,LATAM,SKU-102,2024-05,592,180,577,0,No,Yes,1.0,180.0,749


In [53]:
medical_groupby.sort_index()

,Category,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
0,Infusion Pumps,APAC,SKU-100,2025-08,244,132,245,0,No,Yes,1.0,132.0,945
1,Infusion Pumps,EMEA,SKU-105,2025-11,290,167,275,0,No,Yes,1.0,167.0,459
2,Infusion Pumps,LATAM,SKU-107,2025-07,299,119,285,0,No,Yes,1.0,119.0,223
3,Infusion Pumps,North America,SKU-105,2024-01,294,143,249,0,No,Yes,1.0,143.0,445
4,MRI Coils,APAC,SKU-103,2024-07,297,105,245,5,No,No,1.0,105.0,665
5,MRI Coils,EMEA,SKU-103,2024-11,292,130,278,0,No,Yes,1.0,130.0,653
6,MRI Coils,LATAM,SKU-105,2024-01,267,138,218,0,No,Yes,1.0,138.0,494
7,MRI Coils,North America,SKU-100,2024-05,292,108,247,0,No,Yes,1.0,108.0,900
8,Orthopedic Screws,APAC,SKU-107,2024-01,599,301,524,0,No,Yes,1.0,301.0,229
9,Orthopedic Screws,EMEA,SKU-100,2024-10,587,242,470,0,No,Yes,1.0,242.0,958


In [54]:
medical_groupby.sort_values(by=['Category','Actual_Sales'],ascending=[True,False])

,Category,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
2,Infusion Pumps,LATAM,SKU-107,2025-07,299,119,285,0,No,Yes,1.0,119.0,223
3,Infusion Pumps,North America,SKU-105,2024-01,294,143,249,0,No,Yes,1.0,143.0,445
1,Infusion Pumps,EMEA,SKU-105,2025-11,290,167,275,0,No,Yes,1.0,167.0,459
0,Infusion Pumps,APAC,SKU-100,2025-08,244,132,245,0,No,Yes,1.0,132.0,945
4,MRI Coils,APAC,SKU-103,2024-07,297,105,245,5,No,No,1.0,105.0,665
5,MRI Coils,EMEA,SKU-103,2024-11,292,130,278,0,No,Yes,1.0,130.0,653
7,MRI Coils,North America,SKU-100,2024-05,292,108,247,0,No,Yes,1.0,108.0,900
6,MRI Coils,LATAM,SKU-105,2024-01,267,138,218,0,No,Yes,1.0,138.0,494
8,Orthopedic Screws,APAC,SKU-107,2024-01,599,301,524,0,No,Yes,1.0,301.0,229
11,Orthopedic Screws,North America,SKU-106,2025-01,596,196,508,0,No,Yes,1.0,196.0,334


In [55]:
medical_groupby.head()

,Category,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
0,Infusion Pumps,APAC,SKU-100,2025-08,244,132,245,0,No,Yes,1.0,132.0,945
1,Infusion Pumps,EMEA,SKU-105,2025-11,290,167,275,0,No,Yes,1.0,167.0,459
2,Infusion Pumps,LATAM,SKU-107,2025-07,299,119,285,0,No,Yes,1.0,119.0,223
3,Infusion Pumps,North America,SKU-105,2024-01,294,143,249,0,No,Yes,1.0,143.0,445
4,MRI Coils,APAC,SKU-103,2024-07,297,105,245,5,No,No,1.0,105.0,665


In [56]:
pd.DataFrame({
    'min':medical_groupby.groupby('Category').Actual_Sales.min(),
    'max':medical_groupby.groupby('Category').Actual_Sales.max()
})

,min,max
Category,,
Infusion Pumps,244,299
MRI Coils,267,297
Orthopedic Screws,587,599
Stents (Cardiac),279,297
Surgical Lasers,120,120


In [57]:
medical_groupby.groupby('Category').Actual_Sales.agg([min,max])

,min,max
Category,,
Infusion Pumps,244,299
MRI Coils,267,297
Orthopedic Screws,587,599
Stents (Cardiac),279,297
Surgical Lasers,120,120


In [58]:
medical_groupby.dtypes

Category                     str
Region                       str
Product_ID                   str
Month                        str
Actual_Sales               int64
ERP_Stat_FC                int64
Consensus_FC               int64
Stockout_Days              int64
Promotion_Flag               str
In_Stock_Flag                str
Promo_Uplift_Factor      float64
Bias_Adjusted_Stat_FC    float64
Index_Backwards            int64
dtype: object

In [59]:
medical_data[pd.isnull(medical_data)]

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
996,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
997,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
998,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [60]:
medical_data

,Product_ID,Category,Region,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
0,SKU-100,Orthopedic Screws,APAC,2024-11,116,332,540,15,No,No,1.0,332.0,999
1,SKU-100,Infusion Pumps,North America,2024-07,269,127,291,5,No,No,1.0,127.0,998
2,SKU-100,Orthopedic Screws,LATAM,2025-02,512,258,523,0,Yes,Yes,1.3,335.4,997
3,SKU-100,Orthopedic Screws,APAC,2025-08,474,158,399,0,No,Yes,1.0,158.0,996
4,SKU-100,Orthopedic Screws,APAC,2024-04,563,262,614,0,Yes,Yes,1.3,340.6,995
...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,SKU-109,MRI Coils,North America,2024-05,10,23,54,15,No,No,1.0,23.0,4
996,SKU-109,Infusion Pumps,LATAM,2024-11,12,24,56,15,No,No,1.0,24.0,3
997,SKU-109,Surgical Lasers,APAC,2025-11,0,20,5,0,Yes,Yes,1.3,26.0,2
998,SKU-109,Infusion Pumps,EMEA,2024-03,186,90,156,0,No,Yes,1.0,90.0,1


In [61]:
medical_groupby.rename(columns={'Category':'Product'})

,Product,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
0,Infusion Pumps,APAC,SKU-100,2025-08,244,132,245,0,No,Yes,1.0,132.0,945
1,Infusion Pumps,EMEA,SKU-105,2025-11,290,167,275,0,No,Yes,1.0,167.0,459
2,Infusion Pumps,LATAM,SKU-107,2025-07,299,119,285,0,No,Yes,1.0,119.0,223
3,Infusion Pumps,North America,SKU-105,2024-01,294,143,249,0,No,Yes,1.0,143.0,445
4,MRI Coils,APAC,SKU-103,2024-07,297,105,245,5,No,No,1.0,105.0,665
5,MRI Coils,EMEA,SKU-103,2024-11,292,130,278,0,No,Yes,1.0,130.0,653
6,MRI Coils,LATAM,SKU-105,2024-01,267,138,218,0,No,Yes,1.0,138.0,494
7,MRI Coils,North America,SKU-100,2024-05,292,108,247,0,No,Yes,1.0,108.0,900
8,Orthopedic Screws,APAC,SKU-107,2024-01,599,301,524,0,No,Yes,1.0,301.0,229
9,Orthopedic Screws,EMEA,SKU-100,2024-10,587,242,470,0,No,Yes,1.0,242.0,958


In [62]:
medical_groupby

,Category,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
0,Infusion Pumps,APAC,SKU-100,2025-08,244,132,245,0,No,Yes,1.0,132.0,945
1,Infusion Pumps,EMEA,SKU-105,2025-11,290,167,275,0,No,Yes,1.0,167.0,459
2,Infusion Pumps,LATAM,SKU-107,2025-07,299,119,285,0,No,Yes,1.0,119.0,223
3,Infusion Pumps,North America,SKU-105,2024-01,294,143,249,0,No,Yes,1.0,143.0,445
4,MRI Coils,APAC,SKU-103,2024-07,297,105,245,5,No,No,1.0,105.0,665
5,MRI Coils,EMEA,SKU-103,2024-11,292,130,278,0,No,Yes,1.0,130.0,653
6,MRI Coils,LATAM,SKU-105,2024-01,267,138,218,0,No,Yes,1.0,138.0,494
7,MRI Coils,North America,SKU-100,2024-05,292,108,247,0,No,Yes,1.0,108.0,900
8,Orthopedic Screws,APAC,SKU-107,2024-01,599,301,524,0,No,Yes,1.0,301.0,229
9,Orthopedic Screws,EMEA,SKU-100,2024-10,587,242,470,0,No,Yes,1.0,242.0,958


In [63]:
medical_groupby.rename(index={0:'first',1:'second'})

,Category,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
first,Infusion Pumps,APAC,SKU-100,2025-08,244,132,245,0,No,Yes,1.0,132.0,945
second,Infusion Pumps,EMEA,SKU-105,2025-11,290,167,275,0,No,Yes,1.0,167.0,459
2,Infusion Pumps,LATAM,SKU-107,2025-07,299,119,285,0,No,Yes,1.0,119.0,223
3,Infusion Pumps,North America,SKU-105,2024-01,294,143,249,0,No,Yes,1.0,143.0,445
4,MRI Coils,APAC,SKU-103,2024-07,297,105,245,5,No,No,1.0,105.0,665
5,MRI Coils,EMEA,SKU-103,2024-11,292,130,278,0,No,Yes,1.0,130.0,653
6,MRI Coils,LATAM,SKU-105,2024-01,267,138,218,0,No,Yes,1.0,138.0,494
7,MRI Coils,North America,SKU-100,2024-05,292,108,247,0,No,Yes,1.0,108.0,900
8,Orthopedic Screws,APAC,SKU-107,2024-01,599,301,524,0,No,Yes,1.0,301.0,229
9,Orthopedic Screws,EMEA,SKU-100,2024-10,587,242,470,0,No,Yes,1.0,242.0,958


In [64]:
medical_groupby.rename_axis("indices",axis="rows").rename_axis("company",axis="columns")

company,Category,Region,Product_ID,Month,Actual_Sales,ERP_Stat_FC,Consensus_FC,Stockout_Days,Promotion_Flag,In_Stock_Flag,Promo_Uplift_Factor,Bias_Adjusted_Stat_FC,Index_Backwards
indices,,,,,,,,,,,,,
0,Infusion Pumps,APAC,SKU-100,2025-08,244,132,245,0,No,Yes,1.0,132.0,945
1,Infusion Pumps,EMEA,SKU-105,2025-11,290,167,275,0,No,Yes,1.0,167.0,459
2,Infusion Pumps,LATAM,SKU-107,2025-07,299,119,285,0,No,Yes,1.0,119.0,223
3,Infusion Pumps,North America,SKU-105,2024-01,294,143,249,0,No,Yes,1.0,143.0,445
4,MRI Coils,APAC,SKU-103,2024-07,297,105,245,5,No,No,1.0,105.0,665
5,MRI Coils,EMEA,SKU-103,2024-11,292,130,278,0,No,Yes,1.0,130.0,653
6,MRI Coils,LATAM,SKU-105,2024-01,267,138,218,0,No,Yes,1.0,138.0,494
7,MRI Coils,North America,SKU-100,2024-05,292,108,247,0,No,Yes,1.0,108.0,900
8,Orthopedic Screws,APAC,SKU-107,2024-01,599,301,524,0,No,Yes,1.0,301.0,229


In [65]:
nee=set()

In [66]:
nee

set()

In [67]:
nee.add(4)

In [68]:
nee

{4}

In [70]:
medical_groupby.shape

(20, 13)